switchboard/src/service.rs

```diff
@@ -0,0 +1,1872 @@
+use crate::auth::db::DbAuth;
+use crate::auth::AuthorizationSource;
+use crate::config::ServiceConfig;
+use crate::perms::RunJobOnSupervisor;
+use crate::service::herd::{Herd, HerdError, ReservationError};
+use crate::service::kanban::{Kanban, KanbanError};
+use crate::service::tag::NaiveTagConfig;
+use crate::sql;
+use crate::sql::api_token::TokenError;
+use crate::sql::job::history::SqlJobEvent;
+use crate::sql::job::{SqlExitStatus, SqlFunctionalState, SqlJob, SqlRestartPolicy};
+use axum::extract::ws::WebSocket;
+use chrono::{DateTime, TimeDelta, Utc};
+use futures_util::stream::FuturesOrdered;
+use futures_util::{FutureExt, StreamExt};
+use sqlx::postgres::types::PgInterval;
+use sqlx::{PgExecutor, PgPool, Postgres};
+use std::collections::{BTreeSet, HashSet};
+use std::sync::Arc;
+use thiserror::Error;
+use tokio::sync::mpsc::UnboundedReceiver;
+use tokio::sync::{mpsc, oneshot, watch, Mutex};
+use tokio::task::JoinHandle;
+use tokio::time::{sleep_until, Instant};
+use tracing::{instrument, Instrument};
+use treadmill_rs::api::switchboard::{
+    ExitStatus, ExtendedJobState, JobEvent, JobRequest, JobResult, JobState, JobStatus,
+    SupervisorStatus,
+};
+use treadmill_rs::api::switchboard_supervisor;
+use treadmill_rs::api::switchboard_supervisor::{
+    ImageSpecification, ReportedSupervisorStatus, SupervisorJobEvent,
+};
+use treadmill_rs::connector::RunningJobState;
+use uuid::Uuid;
+
+pub mod herd;
+pub mod kanban;
+mod socket_connection;
+mod tag;
+
+#[derive(Debug, Error)]
+pub enum ServiceError {
+    #[error("Database error: {0}")]
+    Database(sqlx::Error),
+    #[error("Herd error: {0}")]
+    Herd(HerdError),
+    #[error("Reservation error: {0}")]
+    Reservation(ReservationError),
+    #[error("Kanban error: {0}")]
+    Kanban(KanbanError),
+    #[error("Job's tag configuration did not match any currently registered supervisors")]
+    SupervisorMatchError,
+    #[error("Job deleted between match and launch")]
+    NoLongerQueued,
+    #[error("No such job")]
+    NoSuchJob,
+    #[error("Supervisor already connected")]
+    SupervisorAlreadyConnected,
+}
+
+struct State {
+    kanban: Kanban,
+    herd: Herd,
+}
```

Comment:
This core logic of the switchboard lacks a bunch of documentation. Let's sit together some time and walk through these files and add both documentation to the code, and architectural docs to the book.